.github/workflows/validate.yml

```diff
@@ -49,46 +49,52 @@ on:
         required: false
         type: string
 
-env:
-  # We choose a stable ghc version across all os's
-  # which will be used to do the next release
-  GHC_FOR_RELEASE: "9.4.8"
-  # Ideally we should use the version about to be released for hackage tests and benchmarks
-  GHC_FOR_SOLVER_BENCHMARKS: "9.4.8"
-  GHC_FOR_COMPLETE_HACKAGE_TESTS: "9.4.8"
-  COMMON_FLAGS: "-j 2 -v"
+jobs:
 
-  # See https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#hackage-revisions
-  ALLOWNEWER: ${{ github.event.inputs.allow-newer }}
-  CONSTRAINTS: ${{ github.event.inputs.constraints }}
+  config:
+    runs-on: ubuntu-latest
+    # `matrix` can't access `env`, but it can access `outputs` from `needs` jobs.
+    # So we have to "export" some things here as job outputs.
+    #
+    # And GitHub documents that $GITHUB_ENV doesn't persist between jobs, only between
+    # steps. But somehow the `validate` job was getting it anyway, while `build-alpine`
+    # wasn't.
+    #
+    # I give up. Everything is now explicitly (GitHub hasn't invented iteration yet) listed
+    # as an output and (again explicitly) loaded into the environment in the first step of
+    # each job. Thanks, GitHub.
```

Comment:
I don't get the reason for the "each job" part. E.g. if `build-alpine` only needs `GHC_FOR_RELEASE`, why would you initialize all other 6 constants... My general view is that copy-pasting of these blobs should be avoided at all costs: they are as hard to read as they are to maintain.

Reply:
Actually, copy-pasting one big unchanging blob is easier than picking and choosing the pieces I need. Also, some of those things are used inside other actions (notably `cabal-setup`) and I don't want to have to plumb them explicitly when I can stuff them into the environment. As it is, I already have to specify up to 5 parameters to `cabal-setup` that I can't reasonably stuff into the environment.

If GitHub didn't suck, there would be better ways to deal with this mess; I was specifically trying to avoid that big-blob nonsense.

Reply:
Actually, I wonder if GHA's YAML support is good enough that I could use a reference and only write the blob once per workflow.

Reply:
> Actually, copy-pasting one big unchanging blob is easier 

not for reading or understanding what's going on. When I see `GHC_VERSION_FOR_BENCHMARKS` passed to `build-alpine,` I start doubting whether I understand what's going on and what the actual purpose of `build-alpine` is. 

I highly recommend (re)reading the GHC modularity report to realize that these issues are not just for making freshmen CS student life harder, they actually build up and bring you to a point when it's impossible to decouple anything from anything.

Reply:
So you want explicit plumbing everywhere? You maintain the 50-parameter actions, then.

Reply:
Maybe I misunderstood. Does alpine need all these parameters (because of the setup action it depends upon or something else)? In any case, I wanted less parameters, not more.

Reply:
But you're not letting me get rid of them. (These are not parameters. The `with:` sections are parameters. This is the _environment_. The original validate workflow was already using it pervasively, but it was set in `env`. I want that to be shareable across workflows, which `env` is not; someone has to update those blocks in every workflow currently.)

Reply:
Also, this is why I was complaining about `vars` on Matrix the other day: those would have worked fine, minus all this extra rigmarole… but you can't access them in PRs from forks. Naturally. Dear GitHub, variables are _not_ secrets!

Reply:
Sorry for snapping at you; this stuff is just _so_ broken. Go look at how I had to explicitly set `shell` on every single `run` line in the actions for a glaring example. (No, `defaults` isn't supported in actions, of course.) Which is why it has to be a parameter, when otherwise I'd be inclined to put it in the environment (but that's painful when it's coming out of `matrix` because of Windows, where `'bash'` gets you the wrong `bash` — which needs to be documented).

Reply:
I'm also trying to document some of this stuff as I run across it.

Reply:
Perhaps, I was using the word "parameter" in a loose sense to mean env vars. What jumped at me is that the alpine build shouldn't need access to such a big chunk of the environment. Granted, it does today. But that's not great. And having to fight your way through the kilometers of `echo ... >> $GITHUB_ENV` makes the new version even worse. So, I was hoping at least this little detail could be improved. But if not, I concede.

Reply:
I can in fact track down every use of every environment variable through all the actions and reduce it to a minimal set. The problems are:

* the minimal set isn't so obviously minimal, and
* any change to environment variables means repeating this whole tracking operation, because the way jobs fail is things _silently_ expand to the empty string. (In fact, I just fixed a problem with that in the main `validate` step that was only causing a fail later in `dogfooding` because an artifact silently hadn't been uploaded due to a silent version mismatch.)